src/lib/protocols/hamachi.c

```diff
@@ -64,13 +64,15 @@ static void search_hamachi_udp(struct ndpi_detection_module_struct* ndpi_struct,
 
   NDPI_LOG_DBG(ndpi_struct, "search Hamachi over UDP\n");
 
-  /* Session starts with two such packets containing 4 bytes in 
-   * the middle - probably key exchange or something. Skip.
-   */
-  if (packet->payload_packet_len == 76 &&
-      get_u_int64_t(packet->payload, 0) == 0)
-  {
-    return;
+  /* Skip initial 76-byte handshake (relay mode only) */
+  if (flow->packet_counter <= 2 && packet->payload_packet_len == 76) {
+    /* Quick zero-check for first 8 and last 8 bytes */
+    if (get_u_int64_t(packet->payload, 0) != 0 && 
+        get_u_int64_t(packet->payload, 68) != 0)
+    {
+      goto exclude_hamachi;
+    }
+    return; /* Likely Hamachi handshake */
```

Comment:
I think that we should exclude it also when the packet length of the first two packets is not 76
```
if (flow->packet_counter <= 2) {
  if(packet->payload_packet_len == 76 &&
     get_u_int64_t(packet->payload, 0) == 0 && 
     get_u_int64_t(packet->payload, 68) == 0)
    return;  /* Likely Hamachi handshake */
  exclude;
}
```

Reply:
Ok, you convinced me :) let's try to detect mid-flows....

In the end, is that PR still needed? :D @0xA50C1A1